### PR TITLE
ffmpeg: extra configuration options

### LINKED ISSUE #10700: [question] ffmpeg configure codecs, muxers, etc.
Hey,

This isn't a question as such, but it's not really a bug either!

The ffmpeg recipe doesn't currently allow you to explicitly configure the codecs etc. that are included in the build. To replace my manual build I need to be able to restrict what's included.

ffmpeg  supports a large number different codecs, containers, etc. and trying to provide a comprehensive list and maintain parity when new ones are added etc. doesn't seem like it would scale and would add a maintenance burden. 

I'd like to put in a PR to allow these to be configured in some way, and in light of the above, I propose that boolean options for "disable everything" type flags be added, for example `--disable-everything` or `--disable-encoders`, along with free-form options for flags which take a value such as `--enable-encoder=`, it would be left to the autotools build system to complain about any incorrect values passed for these free-form options.  e.g.
```
options = {
# existing options ...
"disable-everything": [True, False],
"disable-encoders": [True, False],
"disable-decoders": [True, False],
"enable-encoder": "ANY",
"enable-decoder": "ANY",
"enable-muxer": "ANY",
"enable-demuxer":, "ANY",
}
```

This would then allow you enable / disable specific codecs by name or create a minimalist build, for example to avoid licencing/patent issues and only allow MP3 decoding:
```
[options]
ffmpeg:disable-everything=True
ffmpeg:enable-decoder="mp3"
ffmpeg:enable-decoder="opus"
ffmpeg:enable-demuxer="mp3"
ffmpeg:enable-demuxer="ogg"
``` 

Another option would be to instead allow a list of codecs etc. to be passed into the recipe, which would reduce the length of the options list in conanfiles, but at the expense of extra parsing logic in the recipe and potentially very long conanfile option lines. Given the trade off I think I'm leaning towards the above example rather than the one below, but I don't feel strongly about this if more people prefer the list approach.
```
[options]
ffmpeg:disable-everything=True
ffmpeg:enable-decoders="mp3, mp3float, opus"
ffmpeg:enable-demuxer="mp3, ogg"
```

Adding this functionality might also be useful for other purposes, such as mentioned in https://github.com/conan-io/conan-center-index/issues/10571, e.g:
```
[options]
ffmpeg:disable-decoders=True
ffmpeg:disable-demuxers=True
# etc
``` 

I'd like to start working on a PR for this, I'm just looking for some feedback on how to add it to the recipe, is this approach sensible or would another approach be preferred? Individual `enable-thing` options or `enable-things` lists? Any other suggestions?

Thanks for reading!

### PR DESCRIPTION
**Pull request by simoncent**
Specify library name and version:  **ffmpeg/5.0**

This PR fixes #10700 by adding options to allow enabling / disabling various ffmpeg features such as encoder, decoders, muxers etc. See the original issue for the discussion around how to surface these options.

The formatter in VS Code seems to have reformatted some of the existing code, but looking a the pep8 guidelines the lines were previously too long anyway, so I guess I've left it. If this is wrong I'm happy to revert those changes,  or split them into a precursor PR for easier review.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
> debconf: delaying package configuration, since apt-utils is not installed

apt-utils is required. As it's a system package, it should be installed separated as another conan package.

**Comment by simoncent** [comment]
I think the build error is more likely `Unknown option "".` since I haven't made any changes that should affect the dependency tree. I thought I was removing empty options, but I must have a small bug, I'll push a fix up soon.

**Comment by simoncent** [comment]
@uilianries it's failing because of a missing prebuilt dependency, what's the best way of dealing with this? Do I need to build the dependency locally and upload it? The only Mac I have is an old 2012 Macbook Pro, it might suffice...

```
ERROR: Missing prebuilt package for 'openssl/1.1.1n'
Use 'conan search openssl/1.1.1n --table=table.html -r=remote' and open the table.html file to see available packages
Or try to build locally from sources with '--build=openssl'
```

**Comment by uilianries** [comment]
@simoncent Try updating to `1.1.1o` (latest). You can't upload anything to Conan Center, only the official is allowed to push packages, for security. If the build fails with 1.1.1o, we can trigger an internal job to build the missing openssl.

**Comment by simoncent** [comment]
>You can't upload anything to Conan Center, only the official is allowed to push packages, for security.
Ahh yeah, that makes sense. I'll try updating.

**Review by SSE4 — DISMISSED** [review]

**Comment by SSE4** [comment]
> The formatter in VS Code seems to have reformatted some of the existing code, but looking a the pep8 guidelines the lines were previously too long anyway, so I guess I've left it. If this is wrong I'm happy to revert those changes, or split them into a precursor PR for easier review.

yes, in general, it's easier for people to review when it contains only relevant changes. otherwise it might be hard to decouple logical changes from styling. not needed to change for this PR, but it would be nice to follow in your next contributions, as it makes process fast(er).

**Comment by simoncent** [comment]
Yeah I usually try and keep whitespace changes separate, but I didn't realise it had happened until I'd already opened the PR, I'll be more careful in future.

**Comment by simoncent** [comment]
Looks like I might need to update some dependencies again.

**Comment by simoncent** [comment]
Conan centre lists `libpng/1.6.37` as the latest already, so I don't think there is anything I can do. Does a new configuration need uploading on your end?

**Comment by paulharris** [comment]
I had this weird "UPSTREAM" failure with another of my recipes, for days - imagemagick.
I "fixed it" today by making a tiny change to the recipe (I added one blank line somewhere) to force the hash to change,
and that forced the CI to build from scratch.

I think yours will fail because it assumes this recipe version has already been built, and so it is trying to pull in the previously-failed builds.

So try adding some whitespace and re-pushing your recipe.

**Comment by simoncent** [comment]
I'll give it a go.

**Comment by uilianries** [comment]
@simoncent you don't submit a new commit to retrigger the CI. You can close this PR, wait 10 seconds, then re-open it again. The interval is needed to Github push the events and the CI process in time.

**Comment by simoncent** [comment]
@uilianries I did it in response to the comment above https://github.com/conan-io/conan-center-index/pull/10877#issuecomment-1141198637

**Comment by simoncent** [comment]
> > The formatter in VS Code seems to have reformatted some of the existing code, but looking a the pep8 guidelines the lines were previously too long anyway, so I guess I've left it. If this is wrong I'm happy to revert those changes, or split them into a precursor PR for easier review.
> 
> Avoid formatting the code. It makes harder to review. Besides that, we no longer use CRT displays, so you can configure to 120 columns anyway.

Yes, I use 120 in other projects, but the pull request template of this repository specifies pep8, which is 80 characters, without stating an exception for 120. Maybe that should be clarified. 

As you quoted, the formatting was accidental and I offered to revert it, but was advised it was fine to leave it. I'm happy to revert it or re-format it at 120 characters if you prefer.

**Comment by uilianries** [comment]
@simoncent you can keep it, just avoid in the future. We don't enforce pep8 or any format convention as Conan is focused to C/C++ developers, and it would restrict even more PRs, as Python is not the main language.

**Comment by simoncent** [comment]
Could that checkbox be removed from the PR template then? It's misleading for newcomers, it's literally the only reason I setup pep8.

**Comment by uilianries** [comment]
> Could that checkbox be removed from the PR template then? It's misleading for newcomers, it's literally the only reason I setup pep8.

Removed not, but we can improve for sure. PEP8 is not only about line-length. Anyway, your complain is totally valid. Thank you for mentioning the checkbox.

File: recipes/ffmpeg/all/conanfile.py [hunk 1]
```diff
@@ -102,6 +137,41 @@ class FFMpegConan(ConanFile):
         "with_audiotoolbox": True,
         "with_videotoolbox": True,
         "with_programs": True,
+        "disable_everything": False,
+        "disable_all_encoders": False,
+        "disable_encoders": "",
+        "enable_encoders": "",
```
Comment (uilianries) [inline recipes/ffmpeg/all/conanfile.py:143]: I would prefer passing `None` for those default empty options, as we usually do for other libraries.

**Review by SSE4 — DISMISSED** [review]

File: recipes/ffmpeg/all/conanfile.py (hunk 1 above)
Comment (simoncent) [reply recipes/ffmpeg/all/conanfile.py:143]: Yeah can do, it'll make the parsing logic slightly more complicated but nothing major.

**Comment by simoncent** [comment]
Is everyone happy with this now? @SSE4 your review was cleared when I pushed a white space change, sorry! @uilianries are you happy with the PR now? Who can provide the 3rd review? Thanks.

**Comment by SSE4** [comment]
> Is everyone happy with this now? @SSE4 your review was cleared when I pushed a white space change, sorry! @uilianries are you happy with the PR now? Who can provide the 3rd review? Thanks.

two reviews are enough now

**Review by SSE4 — APPROVED** [review]

**Comment by simoncent** [comment]
Great, thanks very much!

**Review by uilianries — APPROVED** [review]
LGTM Nice new gamma of options, thanks!

**Comment by SpaceIm** [comment]
Not sure it's a good idea, this kind of design is a mess for package id handling.

**Pull request merged**